rclpy/src/rclpy_common/src/handle.c

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Python.h>
+#include <stddef.h>
+
+#include "rcutils/allocator.h"
+#include "rcutils/strdup.h"
+#include "rcutils/types/rcutils_ret.h"
+
+#include "rclpy_common/handle.h"
+
+/// Creates a Handle object.
+rclpy_handle_t *
+_rclpy_create_handle(void * ptr, rclpy_handle_destructor_t destructor)
+{
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rclpy_handle_t * handle = allocator.zero_allocate(1, sizeof(rclpy_handle_t), allocator.state);
+  if (!handle) {
+    return NULL;
+  }
+
+  handle->ptr = ptr;
+  handle->ref_count++;
+  handle->destructor = destructor;
+
+  return handle;
+}
+
+/// Adds a dependency to a handle.
+/**
+ * The `dependency` handle reference count is incresead.
+ * The `dependent` handle stores `dependency` into its dependencies list.
+ */
+rcutils_ret_t
+_rclpy_handle_add_dependency(rclpy_handle_t * dependent, rclpy_handle_t * dependency)
+{
+  assert(dependent);
+  assert(dependency);
+
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+
+  dependent->num_of_dependencies++;
```

Comment:
@ivanpauno I think this should not be incremented before we're certain the re-allocation succeeded.